lib/std/Progress.zig

```diff
@@ -1,447 +1,1123 @@
-    std.debug.getStderrMutex().lock();
+
+    return start_serialized_len + storage.len;
+}
+
+fn computeRedraw(serialized_buffer: *Serialized.Buffer) []u8 {
+    const serialized = serialize(serialized_buffer);
+
+    // Now we can analyze our copy of the graph without atomics, reconstructing
+    // children lists which do not exist in the canonical data. These are
+    // needed for tree traversal below.
+
+    var children_buffer: [node_storage_buffer_len]Children = undefined;
+    const children = children_buffer[0..serialized.parents.len];
+
+    @memset(children, .{ .child = .none, .sibling = .none });
+
+    for (serialized.parents, 0..) |parent, child_index_usize| {
+        const child_index: Node.Index = @enumFromInt(child_index_usize);
+        assert(parent != .unused);
+        const parent_index = parent.unwrap() orelse continue;
+        const children_node = &children[@intFromEnum(parent_index)];
+        if (children_node.child.unwrap()) |existing_child_index| {
+            const existing_child = &children[@intFromEnum(existing_child_index)];
+            children[@intFromEnum(child_index)].sibling = existing_child.sibling;
+            existing_child.sibling = child_index.toOptional();
+        } else {
+            children_node.child = child_index.toOptional();
+        }
+    }
+
+    // The strategy is: keep the cursor at the end, and then with every redraw:
+    // move cursor to beginning of line, move cursor up N lines, erase to end of screen, write
+
+    var i: usize = 0;
+    const buf = global_progress.draw_buffer;
+
+    buf[i..][0..start_sync.len].* = start_sync.*;
+    i += start_sync.len;
+
+    i = computeClear(buf, i);
+
+    global_progress.accumulated_newline_count = 0;
+    const root_node_index: Node.Index = @enumFromInt(0);
+    i = computeNode(buf, i, serialized, children, root_node_index);
+
+    buf[i..][0..finish_sync.len].* = finish_sync.*;
+    i += finish_sync.len;
+
+    return buf[0..i];
+}
+
+fn computePrefix(
+    buf: []u8,
+    start_i: usize,
+    serialized: Serialized,
+    children: []const Children,
+    node_index: Node.Index,
+) usize {
+    var i = start_i;
+    const parent_index = serialized.parents[@intFromEnum(node_index)].unwrap() orelse return i;
+    if (serialized.parents[@intFromEnum(parent_index)] == .none) return i;
+    if (@intFromEnum(serialized.parents[@intFromEnum(parent_index)]) == 0 and
+        serialized.storage[0].name[0] == 0)
+    {
+        return i;
+    }
+    i = computePrefix(buf, i, serialized, children, parent_index);
+    if (children[@intFromEnum(parent_index)].sibling == .none) {
+        const prefix = "   ";
+        const upper_bound_len = prefix.len + line_upper_bound_len;
+        if (i + upper_bound_len > buf.len) return buf.len;
+        buf[i..][0..prefix.len].* = prefix.*;
+        i += prefix.len;
+    } else {
+        const upper_bound_len = tree_line.len + line_upper_bound_len;
+        if (i + upper_bound_len > buf.len) return buf.len;
```

Comment:
I'm trying to incorporate my Windows API stuff and am confused about what's going on here.

If there is not enough room in the `buf` for (`│  ` + a max length line), then nothing is written to `buf` and the full length of `buf` is returned. Is the `buf.len` return being used as a pseudo-error here?

Reply:
Yes exactly- it triggers the check in `computeNode`:

```zig
    if (i + line_upper_bound_len > buf.len)
        return start_i;
```